hack/tools/tilt-prepare/main.go

```diff
@@ -0,0 +1,520 @@
+	select {
+	case <-wgDone:
+		break
+	case err := <-errors:
+		// cancel all the running tasks
+		cancel()
+		// consumes all the errors from the channel
+		errList := []error{err}
+	Loop:
+		for {
+			select {
+			case err := <-errors:
+				errList = append(errList, err)
+			default:
+				break Loop
+			}
+		}
+		close(errors)
+		return kerrors.NewAggregate(errList)
+	}
+	return nil
+}
+
+// runTask runs a taskFunction taking care of logging elapsed time and reporting Done to the wait group.
+// NOTE: those actions are common to all the tasks, so they are centralized.
+func runTask(ctx context.Context, wg *sync.WaitGroup, prefix string, f taskFunction, errCh chan error) {
+	klog.Infof("[%s] task started\n", prefix)
+	defer func(start time.Time) {
+		klog.Infof("[%s] task completed, elapsed: %s\n", prefix, time.Since(start))
+	}(time.Now())
+
+	defer wg.Done()
+
+	// This is a catch-all in case a task does not cancel properly, then trigger an error
+	// when the error channel has been already closed.
+	defer func() {
+		if r := recover(); r != nil {
+			klog.Errorf("[%s] Recovered from panic: %s", prefix, r)
+		}
+	}()
+
+	// run the actual task.
+	f(ctx, prefix, errCh)
+}
+
+// makeTask generates a task for invoking a make target.
+func makeTask(name string) taskFunction {
+	return func(ctx context.Context, prefix string, errCh chan error) {
+		cmd := exec.CommandContext(ctx, "make", name)
+
+		var stderr bytes.Buffer
+		cmd.Dir = rootPath
+		cmd.Stderr = &stderr
+		if err := cmd.Run(); err != nil {
+			errCh <- errors.Wrapf(err, "[%s] failed to run %s: %s", prefix, cmd.Args, stderr.Bytes())
+		}
+	}
+}
+
+// preLoadImageTask generates a task for pre-loading an image into kind.
+func preLoadImageTask(image string) taskFunction {
+	return func(ctx context.Context, prefix string, errCh chan error) {
+		docker, err := container.NewDockerClient()
+		if err != nil {
+			errCh <- errors.Wrapf(err, "[%s] failed to create docker client", prefix)
+			return
+		}
+
+		if err := docker.PullContainerImageIfNotExists(ctx, image); err != nil {
+			errCh <- errors.Wrapf(err, "[%s] failed to pull %s", prefix, image)
+			return
+		}
+
+		cmd := exec.CommandContext(ctx,
+			"kind",
+			"load",
+			"docker-image",
+			"--name=capi-test", // TODO: make this configurable
+			image,
+		)
+
+		var stderr bytes.Buffer
+		cmd.Dir = rootPath
+		cmd.Stderr = &stderr
+		if err := cmd.Run(); err != nil {
+			errCh <- errors.Wrapf(err, "[%s] failed to run %s: %s", prefix, cmd.Args, stderr.Bytes())
+		}
+	}
+}
+
+// certManagerTask generates a task for installing cert-manager if not already present.
+func certManagerTask() taskFunction {
+	return func(ctx context.Context, prefix string, errCh chan error) {
+		config, err := config.New("")
+		if err != nil {
+			errCh <- errors.Wrapf(err, "[%s] failed create clusterctl config", prefix)
+			return
+		}
+		cluster := cluster.New(cluster.Kubeconfig{}, config)
```

Comment:
I will look into respecting allowed_contexts, but I think we should avoid having two ways for deploying cert-manager
/hold

Reply:
Yup agree. Maybe we can take the context tilt is using and use it here too. Then we're automatically covered.

_speaking as someone who has deployed on prod clusters because of tools which were using current context in the past_

Reply:
fixed

Reply:
Just saw your solution afterwards, looks good